avocado/utils/asset.py

```diff
@@ -0,0 +1,121 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: Red Hat Inc. 2016
+# Author: Amador Pahim <apahim@redhat.com>
+
+"""
+Asset fetcher from multiple locations
+
+"""
+
+import os
+import urlparse
+import logging
+
+from . import crypto
+from . import download
+
+
+log = logging.getLogger('avocado.test')
+
+
+def _check_file(filename, expected_hash):
+    if not os.path.isfile(filename):
+        log.error('Asset %s not found.' % filename)
+        return False
+
+    if expected_hash is None:
+        log.debug('Skipping hash check for %s.' % filename)
+        return True
+
+    file_hash = crypto.hash_file(filename)
+    if file_hash == expected_hash:
+        log.debug('Asset %s verified.' % filename)
+        return True
+    else:
+        log.error('Asset %s corrupted.' % filename)
+        return False
+
+
+def fetch(params, search='asset'):
+    """
+    Tries to fetch a file from multiple locations, trying the cache
+    first and caching when successfully fetched.
+
+    Expected yaml format:
+
+    asset:
+        name: <name> # Not null
+        md5sum: <hash> # Optional. Skips the check if null.
+        cache_dir: <path> # Optional. Default=/var/tmp.
+        location: # list of locations to try.
+        # Supported schemes are: file, http and ftp.
+        - file://<path>
+        - http://<url>
+        - ftp://<url>
+
+    :param params: yaml parsed options.
+    :param search: (optional) search for yaml option with all the asset
+            data. Default is 'asset'.
+    :return: asset path.
+    """
+
+    name = params.get('name', '*/%s/*' % search, default=None)
+    md5sum = params.get('md5sum', '*/%s/*' % search, default=None)
+    cache_dir = params.get('cache_dir', '*/%s/*' % search, default='/var/tmp')
```

Comment:
For the other situations where one wants to set the cache dir, it's advisable to not set the cache dir to a temporary directory (think iso files).